### PR TITLE
DEV-4456: Accept TimeoutError in state match

### PR DESCRIPTION
**Pull request by linaza-siklu**
DEV-4456: Accept TimeoutError in state match:
Handling TimeoutError exception after timeout pass to accept or not accept TimeoutError and move to next_state or raise the TimeoutError.  
- Added boolean attribute "capture_timeout" to the TopLevelModel Match, the attribute is False by default.
- Handle timeout match in the device_model.py
- Added unit tests to input_test.py to test timeout match and pattern unique values.
- changed matches patterns to be unique value.
- Accept TimeoutError only by one match

### CODE REVIEW

File: setup_script/DeviceSetup.py [hunk 1]
```diff
@@ -30,11 +29,21 @@ def run_state(self, state: State) -> bool:
     def _process_input(self, input_model: Input) -> Match:
         regex_list = [match.pattern for match in input_model.matches]
         logger.info(f"Waiting {input_model.timeout} seconds for any of the following patterns: {regex_list}...")
-
+        # Test that only one match accepts TimeoutError
+        regex_matches_timeout = [match.timeout_capture for match in input_model.matches]
```
[the remark this answers is not in the document]
Comment (linaza-siklu) [reply setup_script/DeviceSetup.py]: @shmuelhazan added it to the pydantic model, please see next commit

File: setup_script/DeviceSetup.py [hunk 2]
```diff
@@ -30,11 +29,17 @@ def run_state(self, state: State) -> bool:
...
         logger.info(f"Waiting {input_model.timeout} seconds for any of the following patterns: {regex_list}...")
-
         try:
             found_pattern = self._device_io.read_expect_regex(regex_list, input_model.timeout)
+
         except TimeoutError:
             logger.info("Timed out.")
+            # capture timeout and return match to the next_state
+            for match in input_model.matches:
+                # capture timeout and return match to the next_state
```
Comment (shmuelhazan) [inline setup_script/DeviceSetup.py:39]: Can you add a more detailed comment?

File: setup_script/models/pydantic/device_model.py [hunk 3]
```diff
@@ -49,6 +52,16 @@ class Input(BaseModel):
     # input possible matches list
     matches: List[Match]
 
+    # Test that only one match accepts TimeoutError
+    @validator('matches')
+    def capture_timeout_in_match(cls, v):
+        regex_matches_timeout = [match.capture_timeout for match in v]
+        try:
+            regex_matches_timeout.count(True) > 1
```
Comment (shmuelhazan) [inline setup_script/models/pydantic/device_model.py:60]: Are you sure that this will throw ValidationError?

Comment (linaza-siklu) [reply setup_script/models/pydantic/device_model.py:60]: Yes

**Comment by berezhinskiy** [comment]
@shmuelhazan @linaza-siklu Hi, I have enabled Jenkins PR job for this repo. Please rebase from the master.

File: setup_script/DeviceSetup.py (hunk 2 above)
Comment (linaza-siklu) [reply setup_script/DeviceSetup.py:39]: okay

**Comment by linaza-siklu** [comment]
> @shmuelhazan @linaza-siklu Hi, I have enabled Jenkins PR job for this repo. Please rebase from the master.

Done.

File: setup_script/models/pydantic/device_model.py (hunk 3 above)
Comment (shmuelhazan) [reply setup_script/models/pydantic/device_model.py:60]: Can you explain how it will do so?

File: setup_script/unit-tests/input_test.py [hunk 4]
```diff
@@ -54,5 +59,74 @@ def test_match(self):
...
+            Checks that the read_expect_regex is called when _process_input is called
+            Checks that the correct match is returned from  _process_input
+        :param mock_read: contains the read function
+        :return: assert or success
+        """
+        input_model = Input(timeout=5, matches=[])
+        mock_read.side_effect = TimeoutError
+        # if there is no pattern match and no TIMEOUT pattern then TimeoutError is raised.
+        with pytest.raises(TimeoutError):
+            assert self.device_setup._process_input(input_model) == input_model.matches[1]
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:109]: How will that work? input_model.mathces[1] does not exists.

File: setup_script/unit-tests/input_test.py [hunk 5]
```diff
@@ -54,5 +59,74 @@ def test_match(self):
...
+        input_model = Input(timeout=5,
+                            matches=[Match(capture_timeout=True, pattern='Login1', next_state=FinalState(name='test2'),
+                                           actions=[]),
+                                     Match(pattern='Login1', next_state=FinalState(name='test'),
+                                           actions=[])
+                                     ])
+        mock_read.return_value = input_model.matches[1].pattern
+        # test unique pattern in match, expected Assertion error if patterns aren't unique
+        with pytest.raises(AssertionError):
+            assert self.device_setup._process_input(input_model) == input_model.matches[1]
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:131]: If I understand correctly, pyexpect won't raise AssertionError in this case. We are capturing this statement's failure.

**Review by shmuelhazan — CHANGES_REQUESTED** [review]

**Comment by shmuelhazan** [comment]
If I understand correctly, we are missing unit-tests for validating that there is only one capture_timeput match.

File: setup_script/DeviceSetup.py [hunk 6]
```diff
@@ -30,11 +29,18 @@ def run_state(self, state: State) -> bool:
...
         regex_list = [match.pattern for match in input_model.matches]
         logger.info(f"Waiting {input_model.timeout} seconds for any of the following patterns: {regex_list}...")
-
         try:
             found_pattern = self._device_io.read_expect_regex(regex_list, input_model.timeout)
+
         except TimeoutError:
             logger.info("Timed out.")
+            # find match that accepts TimeoutError in state matches
+            if input_model.matches is not None:
```
Comment (shmuelhazan) [inline setup_script/DeviceSetup.py:38]: is it valid for matches to be none?

File: setup_script/models/pydantic/device_model.py [hunk 7]
```diff
@@ -48,6 +49,15 @@ class Input(BaseModel):
...
     matches: List[Match]
 
+    # Test that only one match accepts TimeoutError
+    @validator('matches')
+    def capture_timeout_in_match(cls, v):
+        regex_matches_timeout = [match.capture_timeout for match in v]
+        if regex_matches_timeout.count(True) <= 1:
+            return v
+        else:
+            raise ValidationError
```
Comment (shmuelhazan) [inline setup_script/models/pydantic/device_model.py:59]: Currently, you are raising a class object. Can you please use `ValidationError("explain the error here")`.

Comment (linaza-siklu) [reply setup_script/models/pydantic/device_model.py:59]: this does not work using ValidationError, instead i used TypeError

File: setup_script/unit-tests/input_test.py (hunk 4 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py]: changed it in the next commit

File: setup_script/unit-tests/input_test.py [hunk 8]
```diff
@@ -52,7 +54,92 @@ def test_match(self):
...
+                                           actions=[])
+                                     ])
+        mock_read.return_value = input_model.matches[0].pattern
+        # test unique pattern in match, expected Assertion error if patterns aren't unique
+        with pytest.raises(AssertionError):
+            assert self.device_setup._process_input(input_model) == input_model.matches[1]
+        _check_call_read_regex(mock_read, input_model)
+
+    @patch('interfaces.AbstractDeviceIO.AbstractDeviceIO.read_expect_regex')
+    def test_process_input_timeout_validation(self, mock_read: Mock):
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:130]: it is unclear for me what this test does (as opposed to other tests)

**Review by shmuelhazan — CHANGES_REQUESTED** [review]

File: setup_script/unit-tests/input_test.py [hunk 9]
```diff
@@ -52,7 +54,92 @@ def test_match(self):
...
+                                                      actions=[])
+                                                ])
+        mock_read.side_effect = [TimeoutError]
+        # no capture to timeout, Timeout exception is expected.
+        with pytest.raises(TimeoutError):
+            assert self.device_setup._process_input(input_model) == input_model.matches[0]
+        _check_call_read_regex(mock_read, input_model)
 
+    @patch('interfaces.AbstractDeviceIO.AbstractDeviceIO.read_expect_regex')
+    def test_process_input_no_match_no_timeout_string(self, mock_read: Mock):
```
[the remark this answers is not in the document]
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:99]: i will remove the second, not sure why i added it but they test the same issue

File: setup_script/unit-tests/input_test.py (hunk 5 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py]: fixed it in the next commit

File: setup_script/DeviceSetup.py (hunk 6 above)
Comment (linaza-siklu) [reply setup_script/DeviceSetup.py]: No it is not valid for matches to be None, matches can only be list of match 
this line was removed

File: setup_script/unit-tests/input_test.py (hunk 8 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py]: see next commit

File: setup_script/models/pydantic/device_model.py [hunk 10]
```diff
@@ -48,6 +54,23 @@ class Input(BaseModel):
     # input possible matches list
     matches: List[Match]
 
+    # Test that only one match accepts TimeoutError
+    @validator('matches')
+    def capture_timeout_in_match(cls, v):
+        # get all capture_timeout values and validate that only one match accepts TimeoutError
+        regex_matches_timeout = [match.capture_timeout for match in v]
+        #  validate that one match accepts TimeoutError
+        if regex_matches_timeout.count(True) <= 1:
```
Comment (shmuelhazan) [inline setup_script/models/pydantic/device_model.py:63]: can you please flip the condition? (e.g. exit on error condition first, and later continiue)

File: setup_script/models/pydantic/device_model.py [hunk 11]
```diff
@@ -48,6 +54,23 @@ class Input(BaseModel):
...
+            # Validate the match patterns are unique patterns
+            regex_list = [match.pattern for match in v]
+            for elem in regex_list:
+                # each pattern in match must be unique, only one appearance
+                if regex_list.count(elem) > 1:
+                    raise pydantic.ValidationError
```
Comment (shmuelhazan) [inline setup_script/models/pydantic/device_model.py:69]: Can you add text in the error?

File: setup_script/unit-tests/input_test.py [hunk 12]
```diff
@@ -52,7 +57,100 @@ def test_match(self):
...
+        mock_read.side_effect = [TimeoutError]
+        # no capture to timeout, Timeout exception is expected.
+        with pytest.raises(TimeoutError):
+            assert self.device_setup._process_input(input_model)
+        _check_call_read_regex(mock_read, input_model)
+
+    @patch('interfaces.AbstractDeviceIO.AbstractDeviceIO.read_expect_regex')
+    def test_process_input_unique_pattern(self, mock_read):
+        """
+            Check that matches pattern are unique for each machine state
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:102]: Is this description correct?

File: setup_script/unit-tests/input_test.py [hunk 13]
```diff
@@ -52,7 +57,100 @@ def test_match(self):
...
+            Checks that the read_expect_regex is not called when _process_input is called
+            Checks that the correct match is returned from  _process_input
+        :param mock_read: contains the read function
+        :return: assert or success
+        """
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:107]: I don't understand what is being tested here. If there won't be an exception here, would the test fail?

**Review by shmuelhazan — CHANGES_REQUESTED** [review]

File: setup_script/unit-tests/input_test.py (hunk 13 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:107]: this test tests that match pattern is unique, here two patterns are equals to 'Login1', so i tested if a validation error is raised in this case, which is what is expected
exception is raised from Validatore in device_model.py

File: setup_script/unit-tests/input_test.py (hunk 12 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:102]: Check that matches patterns are unique for current state machine

File: setup_script/unit-tests/input_test.py [hunk 14]
```diff
@@ -52,7 +57,100 @@ def test_match(self):
...
+        input_model = Input(timeout=1, matches=[Match(capture_timeout=False, pattern='Login',
+                                                      next_state=FinalState(name='test2'),
+                                                      actions=[])
+                                                ])
+        mock_read.side_effect = [TimeoutError]
+        # no capture to timeout, Timeout exception is expected.
+        with pytest.raises(TimeoutError):
+            assert self.device_setup._process_input(input_model)
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:96]: process_input will throw here. I don't understand why do we need to assert on its return value.

File: setup_script/unit-tests/input_test.py (hunk 13 above)
Comment (shmuelhazan) [reply setup_script/unit-tests/input_test.py:107]: @linaza-siklu Will the test fail if the validation won't be correct?

File: setup_script/unit-tests/input_test.py [hunk 15]
```diff
@@ -52,7 +57,100 @@ def test_match(self):
...
+
+    @patch('interfaces.AbstractDeviceIO.AbstractDeviceIO.read_expect_regex')
+    def test_two_matches_accept_timeout(self, mock_read: Mock):
+        """
+            Checks that ValidationError is raised once two or more matches accept TimeoutError
+            Checks that the read_expect_regex is not called when _process_input is called
+        :param mock_read: contains the read function
+        :return: error or success
+        """
+        try:
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:145]: same here. That won't fail in case the code won't throw.

**Review by shmuelhazan — CHANGES_REQUESTED** [review]

File: setup_script/unit-tests/input_test.py (hunk 13 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:107]: yes, if the the input model has two patterns that are the same then the test fails
the validation error is thrown only after reading the yaml file

File: setup_script/unit-tests/input_test.py (hunk 14 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:96]: my mistake, no need for assert , only call process_input

File: setup_script/unit-tests/input_test.py (hunk 15 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:145]: these two tests will fail only if you change input_model, i only added them to test that ValidationError is raised in two cases:
- case 1: two or more matches accepts TimeoutError
- case 2: two or more matches patterns are identical 
They won't fail unless you change the device_model.py (pydantic model)
i don't understand why this is confusing.

File: setup_script/unit-tests/input_test.py [hunk 16]
```diff
@@ -52,7 +56,100 @@ def test_match(self):
...
+        with pytest.raises(ValidationError):
+            input_model = Input(timeout=5,
+                                matches=[
+                                    Match(capture_timeout=False, pattern='Login1', next_state=FinalState(name='test2'),
+                                          actions=[]),
+                                    Match(capture_timeout=True, pattern='Login1', next_state=FinalState(name='test'),
+                                          actions=[])
+                                    ])
+            self.device_setup._process_input(input_model)
+        mock_read.assert_not_called()
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:116]: do we need these lines?

File: setup_script/unit-tests/input_test.py [hunk 17]
```diff
@@ -52,7 +56,100 @@ def test_match(self):
...
+            input_model = Input(timeout=5,
+                                matches=[Match(capture_timeout=True, pattern='Login',
+                                               next_state=FinalState(name='test2'),
+                                               actions=[]),
+                                         Match(capture_timeout=True, pattern='Login1',
+                                               next_state=FinalState(name='test'),
+                                               actions=[])
+                                         ])
+            self.device_setup._process_input(input_model)
+        mock_read.assert_not_called()
```
Comment (shmuelhazan) [inline setup_script/unit-tests/input_test.py:155]: do we need these lines?

**Review by shmuelhazan — CHANGES_REQUESTED** [review]

File: setup_script/unit-tests/input_test.py (hunk 16 above)
Comment (linaza-siklu) [reply setup_script/unit-tests/input_test.py:116]: No, this is only to make sure functions are not called when Input model isn't valid, do you want me to remove these calls?

Comment (shmuelhazan) [reply setup_script/unit-tests/input_test.py:116]: I think that it is better to have the least amount of code in the unit-tests

**Review by shmuelhazan — APPROVED** [review]

**Pull request merged**